pkg/hubble/metrics/dynamic_flow_processor.go

```diff
@@ -0,0 +1,139 @@
+// Copyright Authors of Cilium
+
+package metrics
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+
+	"github.com/cilium/cilium/pkg/hubble/metrics/api"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// DynamicFlowProcessor represents instance of hubble exporter with dynamic
+// configuration reload.
+type DynamicFlowProcessor struct {
+	logger  logrus.FieldLogger
+	watcher *metricConfigWatcher
+	// Protects against deregistering metric handlers while ProcessFlow is executing, or concurrent config reloads.
+	mutex    lock.RWMutex
+	Metrics  []api.NamedHandler
+	registry *prometheus.Registry
+}
+
+// OnDecodedEvent distributes events across all managed exporters.
+func (d *DynamicFlowProcessor) OnDecodedFlow(ctx context.Context, flow *flowpb.Flow) (bool, error) {
+	select {
+	case <-ctx.Done():
+		return false, nil
+	default:
+	}
+
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+
+	var errs error
+	if d.Metrics != nil {
+		errs = api.ExecuteAllProcessFlow(ctx, flow, &d.Metrics)
+	}
+
+	if errs != nil {
+		d.logger.WithError(errs).Error("Failed to ProcessFlow in metrics handler")
+	}
+	return false, errs
+}
+
+// NewDynamicFlowProcessor creates instance of dynamic hubble flow exporter.
+func NewDynamicFlowProcessor(reg *prometheus.Registry, logger logrus.FieldLogger, configFilePath string) *DynamicFlowProcessor {
+	dynamicFlowProcessor := &DynamicFlowProcessor{
+		logger:   logger,
+		registry: reg,
+	}
+	watcher := NewMetricConfigWatcher(configFilePath, dynamicFlowProcessor.onConfigReload)
+	dynamicFlowProcessor.watcher = watcher
+	return dynamicFlowProcessor
+}
+
+func (d *DynamicFlowProcessor) onConfigReload(ctx context.Context, hash uint64, config api.Config) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	var newHandlers []api.NamedHandler
+	metricNames := config.GetMetricNames()
+
+	curHandlerMap := make(map[string]*api.NamedHandler)
+	if d.Metrics != nil {
+		for _, m := range d.Metrics {
+			curHandlerMap[m.Name] = &m
+		}
+
+		configuredMetricNames := make(map[string]*api.MetricConfig)
+		for _, cm := range config.Metrics {
+			configuredMetricNames[cm.Name] = cm
+		}
+		// Unregister handlers not present in the new config.
+		// This needs to happen first to properly check for conflicting plugins later during registration.
+		for _, m := range d.Metrics {
+			if _, ok := configuredMetricNames[m.Name]; !ok {
+				h := curHandlerMap[m.Name]
+				err := h.Handler.Deinit(d.registry)
+				if err != nil {
+					d.logger.WithField("name", m.Name).WithError(err).Error("Deinit failed for handler")
+				}
+				delete(curHandlerMap, m.Name)
+			}
+		}
+	}
+
+	for _, cm := range config.Metrics {
+		// Existing handler matches new config entry:
+		//   no-op, if config unchanged;
+		//   update handler config, if changed.
+		if m, ok := curHandlerMap[cm.Name]; ok {
+			if reflect.DeepEqual(*m.MetricConfig, *cm) {
```

Comment:
I'm not sure if reflect.DeepEqual will work correctly for `*pb.FlowFilter`. You may need to implement a custom comparison function using https://pkg.go.dev/google.golang.org/protobuf/proto#Equal. 

This change needs tests that validates the behavior of the `DynamicFlowProcessor`

Reply:
Just checked that the update path is hit for a new config with only different filter contents (this means the custom comparators aren't needed in the dynamic exporter).

Reply:
The test `TestHandlersUpdatedInDfpOnConfigChange` checks the state of the `DynamicFlowProcessor` for a sequence of config reloads that register and unregister metrics with `assertHandlersInDfp`. I will also add the case I described above (filter modification). What else did you have in mind?

Reply:
Ah I missed that update. I was looking after the first but didn't see it due to how files get ordered in the diffs. Thanks. Will take a look t that.